### ISSUE TITLE
[v6] PropTypes not checked in custom elements ?

### ISSUE THREAD

**Issue by VladShcherbin** [erikras/redux-form#871]
Hey, I've got a custom component and want to check the passed props:

```js
<Field name="title" label="Title" component={Input} />
```

In `Input` component I want to make sure that _label_ prop is passed, but the propTypes does not work in Input element. Am I doing something wrong?

**Comment by erikras** [comment]
> propTypes does not work in Input element

I'm not sure what you mean by that. You can set
```js
Input.propTypes = {
  label: PropTypes.string.isRequired
}
```
and then not pass `label` to `Field` (to be passed on to `Input`) and the propType warning isn't thrown?

Is that what you mean by "does not work"?

**Comment by VladShcherbin** [comment]
@erikras yes, even a simple element test:

```js
import React, { PropTypes } from 'react'

const Input = (props) => (
  <div className="form-group">
    <label htmlFor={props.name}>{props.label}</label>
    <input id={props.name} {...props} />
  </div>
)

Input.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired
}

export default Input
```

This gives the warning:

```js
<Input />
```

This does not:

```js
<Field name="title" component={Input} />
```

**Comment by erikras** [comment]
Very interesting. Will investigate. 🔍

**Comment by erikras** [comment]
Crazy.
## Stateless Functions
```js
const EmailInput = props => <input type="email" {...props}/>
EmailInput.propTypes = { placeholder: React.PropTypes.string.isRequired }
```
This gives a prop types warning:
```js
<Field name="email" component={props =>
  <EmailInput {...props}/>
}/>
```
And this does not:
```js
<Field name="email" component={EmailInput}/>
```
## Component Classes
If you do this
```js
class EmailInput extends React.Component {
  render() {
    return <input type="email" {...this.props}/>
  }
}
EmailInput.propTypes = { placeholder: React.PropTypes.string.isRequired }
```
...it works fine even with the
```js
<Field name="email" component={EmailInput}/>
```
...syntax.
## Conclusion
There must be something that the JSX syntax is doing that is running the stateless function through a prop types check.

If you give `Field` a component class, `redux-form` is calling `React.createFactory()` on it to convert it into a function that can be called with props.

We need a React expert. @gaearon, got a sec?

**Comment by gaearon** [comment]
I don't quite understand. Can you show me an example that doesn't validate PropTypes without Redux Form API? As in, just a standalone example demonstrating the problem.

**Comment by gaearon** [comment]
The problem is here:

```js
isClass(component) ? React.createFactory(component) : component
```

You are bypassing React completely by calling functional components directly. This way their props are not checked, they don't appear in DevTools, and React has no knowledge about them. 

Generally this shouldn't be necessary. Just always use `createFactory`. Or just use `createElement(component, props)` directly because the factory isn't really that useful. Or use JSX which becomes `createElement`, but in this case remember to make the variable pascal case so JSX picks it up.

**Comment by erikras** [comment]
[Here's my fiddle](https://jsfiddle.net/w3Ley8fc/2/) that I was making when you looked at the code yourself. :-)

**Comment by erikras** [comment]
Changing my fiddle code to
```js
const Field = props => {
  const { component, ...rest } = props
  const Component = component
  return <Component {...rest}/>
}
```
is no good, because then it fails with the `React.DOM.input` factory.

```js
const Field = props => {
  const { component, ...rest } = props
  const factory = React.createFactory(component)
  return factory(rest)
}
```
is also no good for the same reason. Is there an `isFactory()` predicate I can use?

**Comment by gaearon** [comment]
My point is you don’t ever need `isClass`. React *knows* how to render stateless components, just let it render them 😉

**Comment by erikras** [comment]
Yes, I understand that, but I also want to be able to accept factories, like `React.DOM.input`, and React doesn't know how to render those as components.

Perhaps I have to provide my own stateless component wrappers for the input factories, e.g.
```js
import { input, textarea, select } from 'redux-form'
...
<Field component={input}/>
```
...but I have some users who are not using JSX and therefore (I think) would find factory support useful.

**Comment by gaearon** [comment]
This is what `Field` should look like in your fiddle:

```js
const Field = props => {
  const { component, ...rest } = props
  return React.createElement(component, rest);
}
```

Factories are not component types. They are just helpers for getting an element. `React.createElement(React.DOM.input)` won’t work because `React.DOM.input` is not a component—I think this is actually correct because your API is supposed to accept a component.

`React.DOM` factories will be removed eventually. Please don’t rely on them. This is just an artifact that is only useful for people who prefer JSX.

**Comment by gaearon** [comment]
In your case, `<Field component='input' />` is all that is necessary for native components. Because a component type can be *either a string or a custom component (function or class)*. These are valid inputs to `createElement`.

**Comment by erikras** [comment]
_Excellent!!_ Cheers.

**Comment by erikras** [comment]
Glad you got to look at this @ooflorent. You're the one not using JSX, right? Nothing about not accepting factories bothers you?

**Comment by ooflorent** [comment]
Well, not really. I'll just wrap `Field` in my codebase to accept them.
Moreover, I may revise my position since a lot of optimizations target JSX.

**Comment by gaearon** [comment]
Since React encourages wrapping with factories at the import site I don't see a problem for factory users either. For native components, they would pass a string. For custom components, they would pass an unwrapped type that they import anyway.

**Comment by erikras** [comment]
Fix published as [`v6.0.0-alpha-7`](https://github.com/erikras/redux-form/releases/tag/v6.0.0-alpha.7).
